File: src/main/java/org/veupathdb/lib/blast/field/TemplateType.java
package org.veupathdb.lib.blast.field;

import com.fasterxml.jackson.databind.JsonNode;
import org.veupathdb.lib.blast.util.JSONConstructor;
import org.veupathdb.lib.blast.util.JSONValue;

public enum TemplateType implements JSONValue
{
  Coding("coding"),
  CodingAndOptimal("coding_and_optimal"),
  Optimal("optimal");

  private final String value;

  TemplateType(String value) {
    this.value = value;
  }

  public String getValue() {
    return value;
  }

  @Override
  public JsonNode toJSON() {
    return JSONConstructor.newText(value);
  }

  public static TemplateType fromString(String value) {
    for (var val : values())
      if (val.getValue().equals(value))
        return val;

    throw new IllegalArgumentException("Unrecognized template_type value \"" + value + "\".");
  }

  public static TemplateType fromJSON(JsonNode js) {
    return fromString(js.textValue());
  }
}

File: src/main/java/org/veupathdb/lib/blast/field/TemplateType.kt
package org.veupathdb.lib.blast.field

import org.veupathdb.lib.blast.util.JSONConstructor.newText
import org.veupathdb.lib.blast.util.JSONValue
import com.fasterxml.jackson.annotation.JsonValue
import com.fasterxml.jackson.databind.JsonNode
import org.veupathdb.lib.blast.util.JSONConstructor
import com.fasterxml.jackson.annotation.JsonCreator
import java.lang.IllegalArgumentException

enum class TemplateType(val value: String) : JSONValue {
  Coding("coding"), CodingAndOptimal("coding_and_optimal"), Optimal("optimal");

  @JsonValue
  override fun toJSON(): JsonNode = newText(value)

  companion object {
    @JvmStatic
    fun fromString(value: String): TemplateType {
      for (`val` in values()) if (`val`.value == value) return `val`
      throw IllegalArgumentException("Unrecognized template_type value \"$value\".")
    }

    @JvmStatic
    @JsonCreator
    fun fromJSON(js: JsonNode): TemplateType {
      return fromString(js.textValue())
    }
  }
}